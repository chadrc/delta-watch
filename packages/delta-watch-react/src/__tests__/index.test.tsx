import 'jest';
import * as React from 'react';
import DeltaWatchReact from '../index';
import * as renderer from 'react-test-renderer';
import advanceTimersByTime = jest.advanceTimersByTime;

jest.useFakeTimers();

describe(`Core`, () => {
  it(`Module loads`, () => {
    expect(DeltaWatchReact).not.toBeTruthy();
  });

  it(`MakeStore and Watch create a renderable component`, () => {
    let watchableStore = DeltaWatchReact.MakeStore({});
    let comp = () => <div/>;
    let Wrapped = watchableStore.Watch()(comp);
    renderer.create(<Wrapped/>);
  });

  it(`Mutating a watchable store causes a wrapped component to re-render`, () => {
    let {Watch, Store} = DeltaWatchReact.MakeStore({
      message: "Hello"
    });

    let renderCount = 0;
    class Text extends React.Component<any, any> {
      render() {
        renderCount++;
        return (
          <p>{this.props.message}</p>
        );
      }
    }

    let Wrapped = Watch((watcher: any) => ({
      message: watcher.message
    }))(Text);

    let component = renderer.create(<Wrapped/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.message = "Delta Watch Render";

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // 1 for initial render, 1 for re-render
    expect(renderCount).toEqual(2);
  });

  it(`Multiple mutations to store only cause one update`, () => {
    let {Watch, Store} = DeltaWatchReact.MakeStore({
      headline: "Multiple Updates",
      message: "No Updates"
    });

    let renderCount = 0;
    class Text extends React.Component<any, any> {
      render() {
        renderCount++;
        return (
          <div>
            <h1>{this.props.headline}</h1>
            <p>{this.props.message}</p>
          </div>
        );
      }
    }

    let Wrapped = Watch((watcher: any) => ({
      headline: watcher.headline,
      message: watcher.message
    }))(Text);

    let component = renderer.create(<Wrapped/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.headline = "Updated";
    Store.Mutator.message = "One Update";

    // Flush current frame of timeouts, happens automatically in browser
    advanceTimersByTime(0);

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // First one will be canceled due to second
    expect(setTimeout).toHaveBeenCalledTimes(2);
    // 1 for initial render, 1 for re-render
    expect(renderCount).toEqual(2);
  });
});