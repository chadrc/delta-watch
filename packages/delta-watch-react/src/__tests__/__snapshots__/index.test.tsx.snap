// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core Multiple mutations to store only cause one update 1`] = `
<div>
  <h1>
    Multiple Updates
  </h1>
  <p>
    No Updates
  </p>
</div>
`;

exports[`Core Multiple mutations to store only cause one update 2`] = `
<div>
  <h1>
    Updated
  </h1>
  <p>
    One Update
  </p>
</div>
`;

exports[`Core Mutating a watchable store causes a wrapped component to re-render 1`] = `
<p>
  Hello
</p>
`;

exports[`Core Mutating a watchable store causes a wrapped component to re-render 2`] = `
<p>
  Delta Watch Render
</p>
`;
